File: pkg/rtweekend/diffuse_light.go
package rtweekend

type DiffuseLight struct {
    emit interface{}
}

func NewDiffuseLight(emit Texture) *DiffuseLight {
    dl := new(DiffuseLight)

    dl.emit = emit

    return dl
}

func NewDiffuseLightFromColor(c Color) *DiffuseLight {
    sc := NewSolidColor(c)

    return &DiffuseLight{*sc}
}

func (dl DiffuseLight) Emitted(u, v float64, p *Point3) *Color {
    return dl.emit.(Texture).Value(u, v, p)
}

func (dl DiffuseLight) Scatter(
    r_in *Ray,
    rec *HitRecord,
    attenuation *Color,
    scattered *Ray,
) bool {
    return false
}

File: src/lib/diffuse_light.rs
use crate::{Color, HitRecord, Point3, Ray};
use crate::{Material, Texture};

pub struct DiffuseLight<T: Texture> {
    pub emit: T,
}

impl<T: Texture> DiffuseLight<T> {
    pub fn new(emit: T) -> Self {
        Self { emit }
    }
}

impl<T: Texture> Material for DiffuseLight<T> {
    fn emitted(&self, u: f64, v: f64, p: &Point3) -> Color {
        self.emit.value(u, v, p)
    }
    fn scatter(&self, _r_in: &Ray, _rec: &HitRecord) -> Option<(Ray, Point3)> {
        None
    }
}
